### PR TITLE
PHP 8.0 and PHP 8.1 support

### PR DESCRIPTION
**Pull request by goba**
In the works :D

### CODE REVIEW

**Comment by goba** [comment]
I have this test code for PHP 7 and 8.1 to try how they handle the namespace tokens differently:

<?php

var_dump(token_get_all('
<?php

namespace Bar\Baz;

\Drupal\boo::scan();
Drupal\boo::scan();
boo::scan();
Drupal\foo();

'));

print T_NAMESPACE . ' -- ';
print T_NS_SEPARATOR	. ' -- ';
print T_PAAMAYIM_NEKUDOTAYIM . ' -- ';
print T_STRING . ' -- ';
print T_NAME_FULLY_QUALIFIED . ' -- ';
print T_NAME_QUALIFIED . ' -- ';
?>

The short of it is PHP 7 parses the top as T_NAMESPACE, whitespace, then T_STRING, T_NS_SEPARATOR, T_STRING. While PHP 8 parses it as T_NAMESPACE, whitespace and T_NAME_QUALIFIED;

As for the invocation in namespace, PHP 7 parses as T_NS_SEPARATOR, T_STRING, T_NS_SEPARATOR, T_STRING followed by T_PAAMAYIM_NEKUDOTAYIM for fully qualified, while PHP 8 parses as T_NAME_FULLY_QUALIFIED followed by T_PAAMAYIM_NEKUDOTAYIM. For not fully qualified, it is T_STRING, T_NS_SEPARATOR, T_STRING  for PHP 7 and T_NAME_QUALIFIED for PHP 8.

So in summary in PHP 7 for a T_STRING you need to look ahead if there is a T_NS_SEPARATOR to know you had a namespace. For PHP 8, it is explicitly identified with its own token. For declarations of namespaces, there is a T_NAME_QUALIFIED instead of T_STRING, etc... when on PHP 8.

**Comment by goba** [comment]
The namespace definition block parsing still needs work. That is the only remaining fail for DMU.

**Comment by goba** [comment]
Based on my debugging Parser::namespaceName() does not end up with the right $node structure with the new constants, the contents of the qualified namespace name does not end up as a child of the node. But I still don't understand the internals of Pharborist enough to tell why its not, even with xdebug.

**Comment by jcnventura** [comment]
Thanks for this @goba. I've given you also repo access, so that I don't block you anymore with this. I'm more than happy to transfer this to another namespace.
